### PR TITLE
feat!: diamond config change + cleanup legacy [CHAIN-682]

### PR DESCRIPTION
**Pull request by justprosh**
## Description
- simplify chain config re Diamond contract upgrade: `diamond_contract_address` instead of `core_contract_address`, `cc_contract_address` and `market_contract_address`
- remove Subnet.resolve with legacy chain-parsing code and implement in connector
- remove unused decider config fields: `network_api_endpoint`, `network_id`, `matcher_address`, `start_block`, `worker_gas`, `wallet_key`
- renamed `pat_id` field to `cu_ids` and changed type to vec (re multi CC update) in ResolveSubnetResult (fixed in fluencelabs/aqua-lib#88)
- remove legacy system services envs: `FLUENCE_ENV_AQUA_IPFS_EXTERNAL_API_MULTIADDR`, `FLUENCE_ENV_AQUA_IPFS_LOCAL_API_MULTIADDR`, `FLUENCE_ENV_CONNECTOR_JOIN_ALL_DEALS`, `FLUENCE_ENV_CONNECTOR_API_ENDPOINT`, `FLUENCE_ENV_CONNECTOR_CONTRACT_ADDRESS`, `FLUENCE_ENV_CONNECTOR_FROM_BLOCK`, `FLUENCE_ENV_CONNECTOR_WALLET_KEY`, `FLUENCE_ENV_DECIDER_IPFS_MULTIADDR`, `FLUENCE_ENV_CONNECTOR_WORKER_GAS`
### Config changes
Before:
```toml
# ...
[system_services.decider]
  decider_period_sec = 30
  worker_period_sec = 30
  worker_ipfs_multiaddr = "/dns4/ipfs.fluence.dev/tcp/5001"
  network_api_endpoint = "rpc.stage.fluence.dev"
  network_id = 123
  matcher_address = "addr1"
  start_block = "latest"
  worker_gas = 1
  wallet_key = "key"

[chain_config]
  http_endpoint = "rpc.stage.fluence.dev"
  core_contract_address = "addr3"
  cc_contract_address = "addr2"
  market_contract_address = "addr1"
  network_id = 123
  wallet_key = "key"
  # ...
```

After:
```toml
# ...
[system_services.decider]
  decider_period_sec = 30
  worker_period_sec = 30
  worker_ipfs_multiaddr = "/dns4/ipfs.fluence.dev/tcp/5001"

[chain_config]
  http_endpoint = "rpc.stage.fluence.dev"
  diamond_contract_address = "addr"
  network_id = 123
  wallet_key = "key"
  # ...
```
## Checklist
- [ ] The code follows the project's coding conventions and style guidelines.
- [ ] All tests related to the changes have passed successfully.
- [ ] Documentation has been updated to reflect the changes (if applicable).
- [ ] All new and existing unit tests have passed.
- [ ] I have self-reviewed my code and ensured its quality.
- [ ] I have added/updated necessary comments to aid understanding.

## Reviewer Checklist
- [ ] Code has been reviewed for quality and adherence to [guidelines](https://doc.rust-lang.org/1.0.0/style/README.html).
- [ ] Tests have been reviewed and are sufficient to validate the changes.
- [ ] Documentation has been reviewed and is up to date.
- [ ] Any questions or concerns have been addressed.

### CODE REVIEW

File: crates/chain-connector/src/types.rs [hunk 1]
```diff
@@ -138,3 +139,46 @@ impl RawTxReceipt {
         }
     }
 }
+
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq)]
+pub struct Worker {
+    pub unit_id: String,
```
Comment (folex) [inline crates/chain-connector/src/types.rs:145]: Seems like most of the engineering team has settled on `cu_id`

File: crates/chain-connector/Cargo.toml [hunk 2]
```diff
@@ -29,6 +29,7 @@ const-hex = { workspace = true }
 serde = { workspace = true }
 async-trait = { workspace = true }
 libipld = { workspace = true }
+libp2p-identity.workspace = true
```
Comment (folex) [inline crates/chain-connector/Cargo.toml:32]:
```suggestion
libp2p-identity { workspace = true }
```

File: crates/chain-connector/src/builtins.rs [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+        CustomService::new(
+            vec![("resolve", make_resolve_subnet_closure(connector.clone()))],
+            None,
+        ),
+    );
+    builtins
+}
+
+fn make_resolve_subnet_closure(connector: Arc<HttpChainConnector>) -> ServiceFunction {
+    ServiceFunction::Immut(Box::new(move |args, _params| {
```
Comment (folex) [inline crates/chain-connector/src/builtins.rs:66]: Maybe DRY? very hard to read this code, and it literally does nothing.

File: crates/chain-connector/src/builtins.rs [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+        let connector = connector.clone();
+        async move { wrap(resolve_subnet_builtin(connector, args).await) }.boxed()
```
Comment (folex) [inline crates/chain-connector/src/builtins.rs:68]: future::ready?

File: crates/chain-connector/Cargo.toml (hunk 2 above)
Comment (justprosh) [reply crates/chain-connector/Cargo.toml:32]: I saw that we started to write it like this in nox `libp2p-identity.workspace` and it's easier to type btw

File: crates/chain-connector/src/builtins.rs (hunk 4 above)
Comment (justprosh) [reply crates/chain-connector/src/builtins.rs:68]: what do you mean?

Comment (folex) [reply crates/chain-connector/src/builtins.rs:68]: https://doc.rust-lang.org/std/future/fn.ready.html

File: crates/chain-connector/src/types.rs [hunk 5]
```diff
@@ -138,3 +139,46 @@ impl RawTxReceipt {
...
+    type Error = libp2p_identity::ParseError;
+    fn try_from(unit: ComputeUnit) -> std::result::Result<Self, Self::Error> {
+        let peer_id = parse_peer_id(&unit.peerId.0)?;
+        let mut worker_id = vec![];
+        if !unit.workerId.is_zero() {
+            worker_id.push(parse_peer_id(&unit.workerId.0)?.to_base58())
+        }
+        let cu_id = unit.id.to_string();
+
+        Ok(Self {
```
Comment (folex) [inline crates/chain-connector/src/types.rs:160]:
```suggestion
        let mut worker_id = vec![];
        if !unit.workerId.is_zero() {
            worker_id.push(parse_peer_id(&unit.workerId.0)?.to_base58())
        }
        let cu_id = unit.id.to_string();
        let peer_id = parse_peer_id(&unit.peerId.0)?;
        
        Ok(Self {
```

Comment (folex) [reply crates/chain-connector/src/types.rs:160]: Also I'd add context: `parse_peer_id(&unit.workerId.0).context("unit.workerId")?` & `parse_peer_id(&unit.peerId.0).context("unit.peerId")?`, so it is possible to differentiate.

File: crates/nox-tests/Cargo.toml [hunk 6]
```diff
@@ -63,3 +65,5 @@ tempfile = { workspace = true }
 jsonrpsee = { workspace = true, features = ["server"] }
 hex = { workspace = true }
 clarity = { workspace = true }
+alloy-sol-types.workspace = true
+alloy-primitives.workspace = true
```
Comment (folex) [inline crates/nox-tests/Cargo.toml:69]: Why not `{ workspace = true }`? Consistency plays very well with multiline editing, search & replace and brings other benefits.

File: crates/nox-tests/tests/builtin.rs [hunk 7]
```diff
@@ -2463,28 +2477,34 @@ async fn subnet_resolve() {
...
     assert_eq!(
-        pats,
+        workers,
         vec![
             (
-                "0x53aadfa1d6cd4c8a18f7eb26bd0b83ca10b664845cd72e2dd871f78b2006f5a7",
-                "12D3KooWN4XNKgu76nwB7iKUXmE4FKCA5Ycak6SbSqLTaWo2nFsQ",
-                vec!["12D3KooWAWdwEujemZN1LQ87bPKSeAvDA1yMGistnL4yF8awuUqV".to_string()],
+                vec![
+                    "0x0000000000000000000000000000000000000000000000000000000000000000"
```
Comment (folex) [inline crates/nox-tests/tests/builtin.rs:2491]: Why is it zeroes here?

File: crates/server-config/src/node_config.rs [hunk 8]
```diff
@@ -329,34 +329,6 @@ impl UnresolvedNodeConfig {
                 _ => {}
             }
         }
-        if let Ok(decider_api_endpoint) = std::env::var("FLUENCE_ENV_CONNECTOR_API_ENDPOINT") {
```
Comment (folex) [inline crates/server-config/src/node_config.rs:332]: 🦾

Comment (folex) [reply crates/server-config/src/node_config.rs:332]: Maybe remove `FLUENCE_ENV_CONNECTOR_JOIN_ALL_DEALS` as well?

And maybe all of these horrifying envs?

File: crates/test-utils/Cargo.toml [hunk 9]
```diff
@@ -17,3 +17,5 @@ maplit = { workspace = true }
 base64 = { workspace = true }
 eyre = { workspace = true }
 cpu-utils = { workspace = true }
+clarity.workspace = true
+server-config.workspace = true
```
Comment (folex) [inline crates/test-utils/Cargo.toml:21]:
```suggestion
clarity = { workspace = true }
server-config = { workspace = true }
```

**Review by folex — APPROVED** [review]
I didn't quite get the tests changes, otherwise LGTM

File: crates/chain-connector/src/builtins.rs [hunk 10]
```diff
@@ -0,0 +1,206 @@
...
+    connector: Arc<HttpChainConnector>,
+    args: Args,
+) -> Result<JValue, JError> {
+    let deal_id: String = Args::next("deal_id", &mut args.function_args.into_iter())?;
+    let deal_id = DealId::from(deal_id);
+
+    let workers: eyre::Result<Vec<Worker>> = try {
+        if !deal_id.is_valid() {
+            Err(eyre::eyre!(
+                "Invalid deal id '{}': invalid length",
```
[the remark this answers is not in the document]
Comment (justprosh) [reply crates/chain-connector/src/builtins.rs:171]: It was implemented in previous version of subnet resolver. I moved it and added todo to make DealID type properly

File: crates/chain-connector/src/connector.rs [hunk 11]
```diff
@@ -519,52 +363,68 @@ impl HttpChainConnector {
...
+    }
+
     fn init_timestamp_params(&self) -> ArrayParams {
         let data = Core::initTimestampCall {}.abi_encode();
         rpc_params![
-            EthCall::to(&data, &self.config.core_contract_address),
+            EthCall::to(&data, &self.config.diamond_contract_address),
             "latest"
         ]
     }
```
[the remark this answers is not in the document]
Comment (justprosh) [reply crates/chain-connector/src/connector.rs:393]: It make sense since we have only one address

File: crates/chain-connector/src/builtins.rs [hunk 12]
```diff
@@ -55,43 +69,20 @@ pub(crate) fn make_connector_builtins(
...
-    ServiceFunction::Immut(Box::new(move |args, params| {
-        let connector = connector.clone();
-        async move { wrap(get_tx_receipts_builtin(connector, args, params).await) }.boxed()
-    }))
-}
-
 async fn get_deals_builtin(
     connector: Arc<HttpChainConnector>,
+    _args: Args,
+
```
Comment (folex) [inline crates/chain-connector/src/builtins.rs:85]:
```suggestion
```

File: crates/chain-connector/src/types.rs [hunk 13]
```diff
@@ -148,14 +149,17 @@ pub struct Worker {
...
-    type Error = libp2p_identity::ParseError;
-    fn try_from(unit: ComputeUnit) -> std::result::Result<Self, Self::Error> {
-        let peer_id = parse_peer_id(&unit.peerId.0)?;
+    type Error = Report;
+    fn try_from(unit: ComputeUnit) -> eyre::Result<Self> {
         let mut worker_id = vec![];
         if !unit.workerId.is_zero() {
-            worker_id.push(parse_peer_id(&unit.workerId.0)?.to_base58())
+            let w_id = parse_peer_id(&unit.workerId.0)
+                .context("unit.workerId")?
```
Comment (folex) [inline crates/chain-connector/src/types.rs:157]: I wonder how these errors would look like 🤔

**Review by folex — APPROVED** [review]
I wond

File: crates/chain-connector/src/types.rs (hunk 13 above)
Comment (justprosh) [reply crates/chain-connector/src/types.rs:157]: btw It's hard to produce this error, it'll fail if it's not 32 bytes

Comment (justprosh) [reply crates/chain-connector/src/types.rs:157]: Eyre doesn't encourage to use it, i will use map_err
With context it looks like `"unit.workerId"`
With `.map_err(|err| eyre!("Failed to parse unit.workerId: {err}"))?` it's  `["Failed to parse unit.workerId: invalid multihash"]`

**Pull request merged**